### DISCUSSION TITLE
Allow readonly domain models take 4

### DISCUSSION THREAD

**Discussion by xenoterracide** [mikro-orm/mikro-orm#3928] [General]
**Is your feature request related to a problem? Please describe.**
Sorry, but I have a new idea on this problem, that is perhaps more acceptable, since it only involves type mutation

My current workaround is to treat your repositories like dao's, and expose a readonly interface. However in my rich models I've noticed my rich methods are also exposed on the where filters and such.

note: `StorefrontProductAggregate` is somewhat readwrite while `StorefrontProduct` is `DeepReadonly<StorefrontProductAggregate>`, this is to avoid the getter/setter issues, whilst protecting the aggregate invariant.

```ts
 constructor(
    @InjectRepository(StorefrontProductAggregate)
    private readonly dao: EntityRepository<DeepWritable<StorefrontProduct>>,
    ) {}

  async findOneOrFail(
    storefrontId: StorefrontId,
    productId: ProductId,
  ): Promise<StorefrontProduct> {
    return await this.dao.findOneOrFail({
      storefrontId,
      productId,
    });
  }
...


**Describe the solution you'd like**

`DeepWriteable` comes from `ts-essentials` I can imagine if you didn't want the dependencies, you could simply copy with attribution.

I should be able to write

```ts
 @InjectRepository(StorefrontProductAggregate)
 private readonly dao: EntityRepository<StorefrontProduct>
```

the filter parameter for "where" should then be based on

```ts
DeepWritable<OmitProperties<StorefrontProduct, Function>>
```

whilst the return should still be the readonly interface of `StorefrontProduct`

---

I might be able to figure out a patch for this either using ts-essentials, or copy-pasta-ing it

**Comment by B4nan** [comment]
I am not sure what are you proposing here.

One thing I can tell you right ahead - the `FilterQuery` and other related types are very complex, and adding one more level of abstraction often breaks the inference (which is imho a TS limitation/bug, but that's what we need to live with). Note that we already omit function types from there:

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/typings.ts#L86

**Reply by xenoterracide** [reply]
simply changing the types inferred by repositories so that every model being write-able and exposed is not required, also so that methods aren't exposed on filters which would never work anyways. The functions do seem to be available according to vscode.

**Reply by B4nan** [reply]
This should help, note that we cant omit getters, they are in fact readonly properties (from TS point of view), which we want to keep queryable.

85b1fc13399a04539f8dcfa31ef12aaef540aa95
